### PR TITLE
[ML] Implement train-test split saving feature in RDataLoader

### PR DESCRIPTION
**Pull request by siliataider**
# This Pull request:

## Changes:

Adds `RDataLoader.save()` to materialise a configured training dataset to disk: the shuffling, splitting and padding a loader performs can be written once then reloaded/distributed and reused instead of being redone on every run.

`save()` drains one full epoch through the existing batch pipeline and fills a `TTree` or `RNTuple` from the batches instead of yielding them. The dataset's config is inherited from the loader.

Columns keep the names given to the loader: a vector column is written as one fixed-size, padded `float` column instead of being exploded into `v1_0 … v1_N`, so a saved file can be fed straight back into a new `RDataLoader` with the same `columns`/`target` arguments.

```python
from ROOT.Experimental.ML import RDataLoader

dl = RDataLoader([df1, df2], columns=["a", "b"], target="label", shuffle=True, set_seed=42)
train, val = dl.train_test_split(test_size=0.2)

train.save("tree", "train.root")  # TTree (default)
val.save("tree", "val.root", output_format="rntuple")  # RNTuple
```

The output format is resolved by a new `RBatchSink` interface that orchestrates the writers:

```mermaid
flowchart LR
  RDF["RDataFrame(s)"] --> CL["RClusterLoader<br>filters, split, shuffle"]
  CL --> BL["RBatchLoader<br>batch_size, drop_remainder"]
  BL --> B["batch (RFlat2DMatrix)"]
  B --> T["as_torch() / as_numpy()"]
  B --> S["save() → RBatchSink::FillBatch()"]
  S --> TT["RTTreeBatchSink<br>TTree::Fill()"]
  S --> RN["RNTupleBatchSink<br>RNTupleWriter::Fill()"]
  TT --> OUT["output file"]
  RN --> OUT
```

> [!NOTE]
> Based on @JazzyJonah's prototype: https://github.com/root-project/root/compare/master...JazzyJonah:root:rdl_save

Commits were assisted-by: Claude Opus 5

### CODE REVIEW

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py [hunk 1]
```diff
@@ -494,6 +496,9 @@ def GetValidationBatch(self) -> Any:
         batch = self.engine.GetValidationBatch()
         return batch if (batch and batch.GetSize() > 0) else None
 
+    def Save(self, treename: str, filename: str, is_training: bool, output_format: str) -> None:
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py]: Not for this PR per se, but I noticed the use of `treename`. We should revisit the data loader codebase and make it more generic so that it doesn't assume the input data format is TTree. For example in this case `dataset_name` should be the better naming choice.

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py [hunk 2]
```diff
@@ -710,6 +716,24 @@ def train_test_split(self, test_size: float = 0.2) -> Tuple[RDataLoader, RDataLo
             RDataLoader._from_internal(self._internal, is_training=False),
         )
 
+    def save(self, treename: str, filename: str | os.PathLike, *, output_format: str = "ttree") -> None:
+        r"""
+        \ingroup Py_ML
+        Write this split to disk as it would be batched for training. Columns are written under
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py]:
```suggestion
        Write this split to disk as it would be loaded for training. Columns are written under
```

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py [hunk 3]
```diff
@@ -710,6 +716,24 @@ def train_test_split(self, test_size: float = 0.2) -> Tuple[RDataLoader, RDataLo
...
+        their original names, a vector column stays one column (now a fixed-size and padded).
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py]:
```suggestion
        Write this split to disk as it would be batched for training. This function preserves the input column names when writing the output dataset. Input columns of collection types keep the same type, but their values will be fixed-sized and padded according to the configuration during loading (e.g. via `max_vec_sizes` and `vec_padding`).
```

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py [hunk 4]
```diff
@@ -710,6 +716,24 @@ def train_test_split(self, test_size: float = 0.2) -> Tuple[RDataLoader, RDataLo
             RDataLoader._from_internal(self._internal, is_training=False),
         )
 
+    def save(self, treename: str, filename: str | os.PathLike, *, output_format: str = "ttree") -> None:
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py]: I'm not sure if we `str` is the right type for `output_format`. Perhaps an `Enum` would be a good choice? The names of the formats would be `TTree` and `RNTuple` so the casing would be important too.

File: bindings/pyroot/pythonizations/test/ml_dataloader.py [hunk 5]
```diff
@@ -7742,5 +7742,211 @@ def test16_vector_padding(self):
...
+
+        df = ROOT.RDataFrame(self.tree_name, self.file_name).Define("v1", "ROOT::VecOps::RVec<int>{ b1,  b1 * 10}")
+
+        dl = ROOT.Experimental.ML.RDataLoader(
+            df,
+            batch_size=5,
+            batches_in_memory=2,
+            columns=["b1", "v1"],
+            max_vec_sizes={"v1": 3},
+            vec_padding=0.0,
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/test/ml_dataloader.py]: Just to make it more meaningful
```suggestion
            vec_padding=42.0,
```

File: bindings/pyroot/pythonizations/test/ml_dataloader.py [hunk 6]
```diff
@@ -7742,5 +7742,211 @@ def test16_vector_padding(self):
...
+        )
+
+        dl.save(self.tree_name, self.out_name)
+
+        saved_b1 = self.read_back(self.out_name, "b1")
+        np.testing.assert_array_equal(np.sort(saved_b1), np.arange(0, 20, 2))
+
+        filter_entries_after = df.AsNumpy(["rdfentry_"])["rdfentry_"]
+
+        self.assertTrue(np.array_equal(filter_entries_before, filter_entries_after))
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/test/ml_dataloader.py]: I'm not sure I understand this test, isn't it always true? Going through the `RDataLoader` shouldn't modify the test of the original `df` at the beginning of a new computation graph run.

File: bindings/pyroot/pythonizations/test/ml_dataloader.py [hunk 7]
```diff
@@ -7742,5 +7742,211 @@ def test16_vector_padding(self):
...
+            drop_remainder=False,
+            load_eager=True,
+        )
+
+        dl.save(self.tree_name, self.out_name)
+
+        np.testing.assert_array_equal(self.read_back(self.out_name, "b1"), np.arange(20))
+        np.testing.assert_array_equal(self.read_back(self.out_name, "b2"), np.arange(1, 21))
+
+
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/test/ml_dataloader.py:7975]: I think we're missing a test of the functionality stated in the PR description
```
... so a saved file can be fed straight back into a new RDataLoader with the same columns/target arguments.
```

File: tree/ml/inc/ROOT/ML/RBatchSink.hxx [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+      for (std::size_t row = 0; row < batch.GetRows(); row++)
+         FillRow(data + row * fRowWidth);
+   }
+
+   /// \brief Flush everything to disk once after the last batch
+   virtual void Commit() = 0;
+};
+
+/// \brief Writes batches into a TTree, as one float leaf per scalar column and one fixed-size
+/// float array leaf per vector column.
```
Comment (vepadulano) [inline tree/ml/inc/ROOT/ML/RBatchSink.hxx]: It would be nice to have a matching output dataset schema for the two sinks, so I would propose to just use `std::vector` as the common type to store collections.

File: tree/ml/inc/ROOT/ML/RBatchSink.hxx [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+#include <stdexcept>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "ROOT/ML/RFlat2DMatrix.hxx"
+#include "ROOT/RNTupleModel.hxx"
+#include "ROOT/RNTupleWriter.hxx"
+#include "TFile.h"
+#include "TTree.h"
```
Comment (vepadulano) [inline tree/ml/inc/ROOT/ML/RBatchSink.hxx]: Given the classes in this header are non-templates, we can separate the implementation in a .cxx and remove all other ROOT includes in this header. This would save a few includes in JITted code (happening when creating the RDataLoaderEngine from Python)

File: tree/ml/CMakeLists.txt [hunk 10]
```diff
@@ -17,4 +18,5 @@ ROOT_STANDARD_LIBRARY_PACKAGE(ROOTMLDataLoader
     src/RSampler.cxx
   DEPENDENCIES
     ROOTDataFrame
+    ROOTNTuple
```
Comment (vepadulano) [inline tree/ml/CMakeLists.txt]: This should be a transitive dependency via ROOTDataFrame

**Review by vepadulano — CHANGES_REQUESTED** [review]
Nice work 🚀 Some comments from my side.

File: bindings/pyroot/pythonizations/test/ml_dataloader.py (hunk 6 above)
Comment (siliataider) [reply bindings/pyroot/pythonizations/test/ml_dataloader.py]: Mirrors checks in pre-existing filtering tests, for ex. https://github.com/root-project/root/blob/d2f93efc6184df1d0771ad82b85698fe53d0bc5e/bindings/pyroot/pythonizations/test/ml_dataloader.py#L496

you're right, not sure why this wouldn't always be true, removed it here

File: bindings/pyroot/pythonizations/test/ml_dataloader.py (hunk 7 above)
Comment (siliataider) [reply bindings/pyroot/pythonizations/test/ml_dataloader.py:7975]: added `test08_round_trip`

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py (hunk 4 above)
Comment (siliataider) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py]: I added an `OutputFormat`, code updated in the PR description, le tme know how it looks

File: bindings/pyroot/pythonizations/test/ml_dataloader.py [hunk 11]
```diff
@@ -7742,5 +7739,239 @@ def test16_vector_padding(self):
...
+    out_name = "dataloader_save_output.root"
+    out_name2 = "dataloader_save_output2.root"
+
+    # Helpers
+    def create_file(self, file_name=None, num_of_entries=20, offset=0):
+        (
+            ROOT.RDataFrame(num_of_entries)
+            .Define("b1", f"(int) rdfentry_ + {offset}")
+            .Define("b2", "b1 + 1")
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/test/ml_dataloader.py:7755]: I'm noticing now that in these tests the input columns are all scalars so we're missing coverage on the vector columns, which should be added especially for `test_08`.

File: tree/ml/inc/ROOT/ML/RBatchSink.hxx [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+   {
+      for (const auto &col : fLayout)
+         fRowWidth += col.fWidth;
+   }
+
+   /// \brief Copy one row of \p fRowWidth floats into the writer's buffers and write the entry
+   virtual void FillRow(const float *row) = 0;
+
+public:
+   virtual ~RBatchSink() = default;
```
Comment (vepadulano) [inline tree/ml/inc/ROOT/ML/RBatchSink.hxx:66]: Use rule of five here.

Comment (vepadulano) [reply tree/ml/inc/ROOT/ML/RBatchSink.hxx:66]: And in derived classes.

File: tree/ml/inc/ROOT/ML/RBatchSink.hxx [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+   /// \brief Write every row of \p batch.
+   void FillBatch(const RFlat2DMatrix &batch)
```
Comment (vepadulano) [inline tree/ml/inc/ROOT/ML/RBatchSink.hxx:69]: Since there is an implementation file anyway, might as well only forward declare RFlat2DMatrix here and get rid of the last ROOT include.

**Review by vepadulano — CHANGES_REQUESTED** [review]

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py (hunk 4 above)
Comment (siliataider) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_ml_dataloader.py]: So we reverted back to strings for `output_format` ("ttree"/"rntuple") instead of the enum. Enums are cleaner in general but probably not the first syntax that would come to a Python user's mind for this type of interface, other "similar" interfaces use strings like `matplotlib.savefig(format="png", ...)` and Hugging face's `load_dataset("json",...)`

**Review by vepadulano — APPROVED** [review]
LGTM, thank you!
